File: EBNFParser/CSharp/LanguageTest/SelfExaminationForEBNF/Parser.cs
using System.Collections;
using System.Collections.Generic;
using CSharp.ObjectRegex;
using System;
namespace CSharp.LanguageTest.SelfExaminationForEBNF
{
    public class Parser
    {
        public Ast Expr ;
        public Ast Or;
        public Ast SPLIT;
        public Ast Atom;
        public Ast AtomExpr;
        public Ast Eq;
        public Ast Stmt;
        public Ast Trailer;
        public Dictionary<string, Ast> compile_closure;
        public Parser(Ast Expr, Ast Or, Ast SPLIT,
                      Ast Atom, Ast AtomExpr,
                      Ast Eq,   Ast Stmt,
                      Ast Trailer,
                      Dictionary<string, Ast> compile_closure
                      ){
                this.Expr = Expr;
                this.Or   = Or;
                this.SPLIT= SPLIT;
                this.Atom = Atom;
                this.Eq   = Eq;
                this.AtomExpr=AtomExpr;
                this.Stmt = Stmt;
                this.Trailer = Trailer;
                this.compile_closure = compile_closure;

        }
        
        public static Parser GenParser(){
            var compile_closure = new Dictionary<string, Ast>();
            var Expr = new Ast(
                compile_closure:ref compile_closure,
                name:"Expr",
                ebnf: new BaseAst[]{
                    new Seq(
                        compile_closure:ref compile_closure,
                        atleast:0,
                        name:"Or+",
                        ebnf:new BaseAst[]{
                            new LazyDef("Or")
                        }
                    )
                }
            );

            var Or = new Ast(
                compile_closure:ref compile_closure,
                name:"Or",
                ebnf:new BaseAst[]{
                    new LazyDef("AtomExpr"),
                    new Seq(
                        compile_closure:ref compile_closure,
                        atleast:0,
                        name:"('|' AtomExpr)*",
                        ebnf:new BaseAst[]{
                            DefualtToken.OrSign,
                            new LazyDef("AtomExpr")
                        }
                    )
                }
            );
            var AtomExpr = new Ast(
                compile_closure:ref compile_closure,
                name:"AtomExpr",
                ebnf:new BaseAst[]{
                    new LazyDef("Atom"),
                    new LazyDef("Trailer")
                }
            );
            var Trailer = new Seq(
                compile_closure:ref compile_closure,
                name:"Trailer",
                atleast:0,
                ebnf:new BaseAst[][]{
                    new BaseAst[]{
                        DefualtToken.SeqStar
                    }, 
                    new BaseAst[]{
                        DefualtToken.SeqPlus
                    },
                    new BaseAst[]{
                        DefualtToken.LBB,
                        new Seq(
                            compile_closure:ref compile_closure,
                            atleast:1,
                            atmost:2,
                            name:"Number{1 2}",
                            ebnf: new BaseAst[]{
                                DefualtToken.Number
                            }),
                        DefualtToken.RBB
                    }
                }
            );
            var Atom = new Ast(
                compile_closure:ref compile_closure,
                name:"Atom",
                ebnf:new BaseAst[][]{
                    new BaseAst[]{
                        DefualtToken.Name
                    },
                    new BaseAst[]{
                        DefualtToken.Str
                    },
                    new BaseAst[]{
                        DefualtToken.LB,
                        new LazyDef("Expr"),
                        DefualtToken.RB
                    },
                    new BaseAst[]{
                        DefualtToken.LP,
                        new LazyDef("Expr"),
                        DefualtToken.RP
                    }
                }
            );
            var Eq = new Ast(
                compile_closure:ref compile_closure,
                name:"Eq",
                ebnf:new BaseAst[]{
                    DefualtToken.Name,
                    DefualtToken.Def,
                    new LazyDef("Expr")
                }
            );
            var Stmt = new Ast(
                compile_closure:ref compile_closure,
                name:"Stmt",
                ebnf:new BaseAst[]{
                    new Seq(
                        compile_closure:ref compile_closure,
                        name:"Eq*",
                        atleast:0,
                        ebnf:new BaseAst[]{
                            new LazyDef("SPLIT"),
                            new LazyDef("Eq"),
                            new LazyDef("SPLIT"),
                        }
                    )
                }
            );
            var SPLIT = new Seq(
                compile_closure: ref compile_closure,
                name:"SPLIT",
                atleast:0,
                ebnf:new BaseAst[]{
                    DefualtToken.NEWLINE
                }
            );
            var namestore = new HashSet<string>();
            SPLIT.compile(ref namestore);
            Stmt.compile(ref namestore);
                
            return new Parser(
                Expr:Expr,
                Atom:Atom,
                Eq:Eq,
                Or:Or,
                Stmt:Stmt,
                AtomExpr:AtomExpr,
                Trailer:Trailer,
                SPLIT:SPLIT,
                compile_closure:compile_closure
            );

        }
        
        
         
        
    }
}

File: Python/Misakawa/Bootstrap/Parser.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Tue Sep 19 17:16:02 2017

@author: misakawa
"""

from ..ObjectRegex.Node import Ref, AstParser, SeqParser, LiteralParser
lit = LiteralParser 

Str    = lit("[RK]{0,1}'[\w|\W]*?'", name = 'Str')
Name   = lit('[a-zA-Z_\u4e00-\u9fa5][a-zA-Z0-9\u4e00-\u9fa5\.]*', name = 'Name')
Number = lit('\d+',name = 'Number')

NEWLINE= lit('\n', name = 'NEWLINE')

LBB = lit.Eliteral('{', name = 'LBB')
LB  = lit.Eliteral('[', name = 'LB')
LP  = lit.Eliteral('(', name = 'LP')
RBB = lit.Eliteral('}', name = 'RBB')
RB  = lit.Eliteral(']', name = 'RB')
RP  = lit.Eliteral(')', name = 'RP')

SeqStar = lit.Eliteral('*', name = 'SeqStar')
SeqPlus = lit.Eliteral('+', name = 'SeqPlus')

Def     = lit.Eliteral('::=', name = 'Def')
LitDef  = lit.Eliteral(':=', name = 'LitDef')
OrSign  = lit.Eliteral("|",   name = 'OrSign')

Throw   = lit.Eliteral('Throw', name = 'Throw') 
using   = lit.Eliteral('using', name = 'usingSign')
Codes   = lit('\{\{[\w\W]+?\}\}', name = 'Codes')

namespace     = globals()
recurSearcher = set()

Using = AstParser(
        [using, SeqParser([Name],[Codes], atmost = 1)],
        name = 'Using')

Expr = AstParser(
    [Ref('Or'),
     SeqParser([OrSign, Ref('Or')],
               atleast=0)
     ],
    name = 'Expr')
Or = AstParser(
    [SeqParser([Ref('AtomExpr')],
               atleast=1)
     ],
    name = 'Or')

AtomExpr =  AstParser(
    [Ref('Atom'), SeqParser([Ref('Trailer')])],
    name = 'AtomExpr')

Atom = AstParser(
    [Str],
    [Name],
    [LP, Ref("Expr"), RP],
    [LB, Ref("Expr"), RB],
    name = 'Atom')

Equals = AstParser(
    [Name, LitDef, SeqParser([Str], atleast=1)],
    [Name, SeqParser([Throw,SeqParser([Name])],atmost = 1),Def, Ref("Expr")],
    name = 'Equals')


Trailer = AstParser(
    [SeqStar],
    [SeqPlus],
    [LBB, SeqParser(
        [Number],
        atleast=1,
        atmost =2
        ),           RBB],
    name = 'Trailer')

Stmt  = AstParser(
            [SeqParser([Using], atmost = 1),
             SeqParser([
                        SeqParser([NEWLINE]),
                        SeqParser([Ref('Equals')]),
                        SeqParser([NEWLINE])
                       ]),
            ],
            name = 'Stmt', toIgnore = {'NEWLINE'})

Stmt.compile(namespace, recurSearcher)


import re as _re
def _escape(*strs):
    return '|'.join([_re.escape(str) for str in strs])
token = _re.compile('|'.join(
        [
        "[RK]{0,1}'[\w\W]*?'",
        '\{\{[\w\W]+?\}\}',
        _escape('|',
                '{',
                '}',
                '[',
                ']',
                '(',
                ')',
                '+',
                '*',
                ':=',
                '::=',
                '.'),
        "[a-zA-Z_\u4e00-\u9fa5][a-zA-Z0-9\u4e00-\u9fa5\.]*",
        "\d+","\n",
        ]
        ))








